### PR TITLE
feat: initial setup

### PR DESCRIPTION
**Pull request by quangpdt**
References https://github.com/as24ch/au-docs/discussions/10

## Motivation and context

- Create a lean custom ESLint config for Linter, also add an initial setup for formatter using Prettier.
- Thanks to the original repo from CarForYou https://github.com/carforyou/carforyou-eslint-config-pkg

## How to test

- Simply run command `npm run test`.

```
info Run bad test cases with config from ./smg-au/eslint-config-pkg/src/__tests__/.eslintrc.js
info Expect to see 13 ESLint errors
info ESLint rules passed
info Run good test cases with config from ./smg-au/eslint-config-pkg/src/__tests__/.eslintrc.js
info Expect to see 0 ESLint errors
info ESLint rules passed
info Run bad test cases with config from ./smg-au/eslint-config-pkg/src/__tests__/.eslintrc.react.js
info Expect to see 21 ESLint errors
info ESLint rules passed
info Run good test cases with config from ./smg-au/eslint-config-pkg/src/__tests__/.eslintrc.react.js
info Expect to see 0 ESLint errors
info ESLint rules passed
info Run bad test cases with config from ./smg-au/eslint-config-pkg/src/__tests__/.eslintrc.next.js
info Expect to see 19 ESLint errors
info ESLint rules passed
info Run good test cases with config from ./smg-au/eslint-config-pkg/src/__tests__/.eslintrc.next.js
info Expect to see 0 ESLint errors
info ESLint rules passed
```

### CODE REVIEW

**Comment by quangpdt** [comment]
Question from @masone :

> is this supposed to be in `fe-eslint-config-pkg`? like.. is the `fe` a prefix we agreed upon? I don't care which way around, but in any case, please keep it consistent :)

I have never heard about this agreement before, have we? Following the `example-pkg` created by @Averethel, there will be no prefix `fe`. But if we do have that convention, I can change the package name and also the repo name

File: src/index.js [hunk 1]
```diff
@@ -0,0 +1,113 @@
+// eslint-disable-next-line import/no-internal-modules
+require('@rushstack/eslint-patch/modern-module-resolution');
```
Comment (quangpdt) [inline src/index.js:2]: From @masone:

> this looks like a workaround for me that I'd like to prevent. can you elaborate on why it's needed?

This actually a workaround, written by Microsoft and have been widely used. Its purpose is giving ability to create a ESLint custom config without  imposing peer dependencies from consumer package:

> This patch is a workaround for a longstanding [ESLint feature request](https://github.com/eslint/eslint/issues/3458) that would allow a shared ESLint config to bring along its own plugins, rather than imposing peer dependencies on every consumer of the config. In a monorepo scenario, this enables your lint setup to be consolidated in a single NPM package.

Read more at: https://www.npmjs.com/package/@rushstack/eslint-patch

File: README.md [hunk 2]
```diff
@@ -1,42 +1,49 @@
...
+  }
+```
+
+- For plain TypeScript project, use the default one: `"@smg-automotive/eslint-config"`
+- For React project, use `"@smg-automotive/eslint-config/react"`
+- For Next.js project, use `"@smg-automotive/eslint-config/next"`
+
+### EditorConfig configuration
+
+Create/Update your `.editorconfig` by [this content](https://github.com/smg-automotive/eslint-config-pkg/blob/main/.editorconfig).
```
Comment (quangpdt) [inline README.md:39]: From @masone 

> Is there some sort of postinstall hook in npm where we could copy the file for the user of the library?

This is a base agreement for `.editorconfig`. If we all agree, we can update the content of this file in `example-pkg`, and other projects can start using it right after repository is created

File: src/index.js (hunk 1 above)
Comment (Averethel) [reply src/index.js:2]: @masone We have two approaches here:
- either the lint configuration bundles all the needed plugins. This way we can ensure consistency but we need the patch
- we define all the plugins as peer deps. This gives us a bit more flexibility but then it's harder to enforce using the same version of the plugin across different projects

File: README.md (hunk 2 above)
Comment (Averethel) [reply README.md:39]: @masone I think there are capabilities for either pos-tinstall messages or actions. I wouldn't go in that direction, though. I wouldn't even add `.editorconfig` to the example repos, since this might overwrite personal settings.
One can set `.editorconfig` on the workspace level, instructing how to do that should be enough.

File: package.json [hunk 3]
```diff
@@ -1,61 +1,58 @@
...
+  },
+  "exports": {
+    "./prettier": "./src/prettier.js",
+    "./next": "./src/next.js",
+    "./react": "./src/react.js",
+    ".": "./src/index.js"
+  },
+  "files": [
+    "src/**/*"
+  ]
```
Comment (Averethel) [inline package.json:57]: Nitpicking alert.
Can we have these grouped with all package related config at the beginning of the package.json?

File: tsconfig.json [hunk 4]
```diff
@@ -1,19 +1,11 @@
 {
```
Comment (Averethel) [inline tsconfig.json:1]: Do we still need tsconfig if we're having this as plain js?

**Review by Averethel — APPROVED** [review]

File: tsconfig.json (hunk 4 above)
Comment (quangpdt) [reply tsconfig.json:1]: The`tsconfig.json` is used for running test cases for TypeScript files. I will move it to more appropriate place,`__tests__` dir

**Comment by Averethel** [comment]
Really cool. Once this is merged and we have it released, please apply the config in the example package too

File: .editorconfig [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+[*]
+charset = utf-8
+indent_style = space
+indent_size = 2
+insert_final_newline = true
+trim_trailing_whitespace = true
+
+[*.{md,mdx}]
+max_line_length = off
+trim_trailing_whitespace = false
```
Comment (quangpdt) [inline .editorconfig:12]: From @Averethel :

> Does one need trailing whitespace in markdown?

From @lkappeler :

> trailing whitespaces are used for line breaks / new paragraph indicators in markdown that's why that are normally allowed

File: .circleci/config.yml [hunk 6]
```diff
@@ -141,7 +125,6 @@ workflows:
           requires:
             - lint
             - unit
-            - typecheck
```
Comment (masone) [inline .circleci/config.yml:144]: did we miss that in the example pkg? if yes, can you please add it there too?

File: src/index.js (hunk 1 above)
Comment (masone) [reply src/index.js:2]: Thanks, I can follow on a highl-level. Sounds like you're following an industry standard and that's cool!

File: .editorconfig [hunk 7]
```diff
@@ -0,0 +1,12 @@
+root = true
+
+[*]
+charset = utf-8
+indent_style = space
+indent_size = 2
+insert_final_newline = true
+trim_trailing_whitespace = true
+
+[*.{md,mdx}]
```
Comment (masone) [inline .editorconfig:10]: I would treat all files the same for the sake of simplicity. no max length will make it incredibly hard to parse long text. I'd be totally in on enforcing that

File: package.json [hunk 8]
```diff
@@ -1,61 +1,58 @@
...
     "eslint-plugin-react-hooks": "4.5.0",
-    "eslint-plugin-testing-library": "5.3.1",
+    "eslint-plugin-testing-library": "5.3.1"
+  },
+  "devDependencies": {
+    "chalk": "4.1.2",
     "jest": "28.0.3",
-    "jest-environment-jsdom": "28.0.2",
     "prettier": "2.6.2",
+    "prop-types": "15.8.1",
```
Comment (masone) [inline package.json:52]: can we work around including this package just for the sake of running a test?

**Review by masone — APPROVED** [review]
I'm confused - why are we repeating my questions through Quangs comments. I can't find mine anymore 🙈 

Anyway, I like this iteration a ton - it's much simpler. Nothing truly blocking from my side, but I would appreciate a quick judgement of the comments before merging

**Comment by Averethel** [comment]
> I'm confused - why are we repeating my questions through Quangs comments. I can't find mine anymore 🙈

We closed #1 and replaced it with this PR due to a branch name that included a slash and prevented a pre-release

File: .circleci/config.yml (hunk 6 above)
Comment (Averethel) [reply .circleci/config.yml:144]: It's removed because this is plain javascript

Comment (masone) [reply .circleci/config.yml:144]: ah removed - I'm sorry, I mixed it up with an addition 🙈

**Comment by masone** [comment]
Also - really nice collabo @quangpdt @Averethel

File: README.md (hunk 2 above)
Comment (Averethel) [reply README.md:39]: It seems like we interpret [discussion in the RFC](https://github.com/smg-automotive/au-docs/discussions/10#discussioncomment-2684546) in different ways. I understood that we would NOT include it in the templates but share it through the package and everyone can configure it/extend their own as they see fit.
Was I off?

File: .editorconfig (hunk 7 above)
Comment (Averethel) [reply .editorconfig:10]: That can really break markdown files. Enabling line wrap should make it easy to follow the text.

File: package.json [hunk 9]
```diff
@@ -1,61 +1,58 @@
...
+    ".": "./src/index.js"
+  },
+  "files": [
+    "src/**/*"
+  ],
   "scripts": {
     "version": "npm run build",
-    "build": "rimraf dist && rollup -c",
-    "lint": "eslint --ext ts,js,tsx,jsx,json .",
+    "build": "echo Success",
```
[the remark this answers is not in the document]
Comment (Averethel) [reply package.json:19]: It could be that semantic release relies on this. But maybe you can make `version` a noop instead?

**Comment by Averethel** [comment]
I only advised/consulted here, full credit goest to @quangpdt. Really nice job 🙏🏻

File: README.md (hunk 2 above)
Comment (Averethel) [reply README.md:39]: Jain, one can still have a broader config for things they like, and since they can be repo specific I'd `.gitignore` this on other repos to allow customisation. But that's my personal preference - if it's easier for the team to have it tin the templates - let's roll with this.

File: .editorconfig (hunk 7 above)
Comment (quangpdt) [reply .editorconfig:10]: Markdown syntax relies a lot on line break. I would suggest we keep this option as it is

File: package.json (hunk 8 above)
Comment (quangpdt) [reply package.json:52]: The `prop-types` was installed as a devDependencies, so it will not affect to the consumer packages anyway

File: package.json (hunk 9 above)
[the remark this answers is not in the document]
Comment (Averethel) [reply package.json:19]: yes, there's nothing to build so no need to run the job and consume resources

**Comment by Averethel** [comment]
@quangpdt can you take a look at the configuration of `import/no-internal-modules`? It currently prevents me from doing relative imports

**Comment by Averethel** [comment]
It also looks like eslint is not provided, when I just install the package, add the config and try to lint I am getting `command not found`. If this is the case and the project needs to provide the eslint - let's make it a peerDependency

**Comment by quangpdt** [comment]
> It also looks like eslint is not provided, when I just install the package, add the config and try to lint I am getting `command not found`. If this is the case and the project needs to provide the eslint - let's make it a peerDependency

Using ESLint Patch from Rushstack allows us not to install ESLint as Dev dependency anymore. I checked your `package.json` it might have a conflict problem, the only dependency should be kept in `devDepdendencies` is: `"eslint-plugin-storybook": "0.5.11",`.

I removed all other plugins and it works so far, however, there're too many NPM error logs regarding to Storybook awaiting for your fix 🥲

**Comment by quangpdt** [comment]
> @quangpdt can you take a look at the configuration of import/no-internal-modules? It currently prevents me from doing relative imports

@Averethel. I understand the problem. From perspective of `import/no-internal-modules`, each child directory of `src` should serve for a single purpose and all public modules should be exported as a single point, `index.ts`. Personally, I found it's quite too strict to use for development and only should be applied when importing from Dependencies, there's still [open discussion](https://github.com/import-js/eslint-plugin-import/issues/1430) regarding this topic.

To get over this issue, I would recommend to update `tsconfig.json` and your `.eslintrc.js` for minor customization:

```json
{
  "compilerOptions": {
    "baseUrl": "./src",
    "paths": {
      "~/*": ["./*"]
    }
  },
}
```

Then, allow internal modules could be imported within our app scope in `.eslintrc.js` with custom rule:

```js
'import/no-internal-modules': [
  'error',
  {
    allow: ['~/**/*'],
  },
],
```

For Next.js application, we'll need to do some more tweaks, but if we all agree with this, I can re-adapt to the ESLint config so you don't need to customize the `.eslintrc.js` anymore. My idea is updating rule `import/no-internal-modules` with some exceptions that can cover all kind of projects (TypeScript, React, Next.js)

```js
'import/no-internal-modules': [
  'error',
  {
    allow: ['~/**/*', '@pages/**/*', , '@components/**/*'],
  },
],
```

File: package.json (hunk 8 above)
Comment (masone) [reply package.json:49]: it's still silly/unnecssary. if we can write that test without the dep, let's please do so. even just a dev dependency will want to be updated and will complicate dependency resolution. no code is faster than no code :)

Comment (quangpdt) [reply package.json]: I found the way to run the test for `react/prop-types` without this dev dependency. Removed it from the list already.

**Comment by Averethel** [comment]
> > It also looks like eslint is not provided, when I just install the package, add the config and try to lint I am getting `command not found`. If this is the case and the project needs to provide the eslint - let's make it a peerDependency
> 
> @Averethel
> 
> Using ESLint Patch from Rushstack allows us not to install ESLint as Dev dependency anymore. I checked your `package.json` it might have a conflict problem, the only dependency should be kept in `devDepdendencies` is: `"eslint-plugin-storybook": "0.5.11",`.
> 
> Btw, could you install latest version of `@smg-automotive/eslint-config`, I have just added the [`eslint-import-resolver-typescript`](https://www.npmjs.com/package/eslint-import-resolver-typescript) to resolve all custom path defined in `tsconfig.json`

**Comment by Averethel** [comment]
> Using ESLint Patch from Rushstack allows us not to install ESLint as Dev dependency anymore. I checked your `package.json` it might have a conflict problem, the only dependency should be kept in `devDepdendencies` is: `"eslint-plugin-storybook": "0.5.11",`.

Yup the issue was on my end. For posterity:

This could happen because eslint was a dependency and got removed. A workaround would be to remove `node_modules` and `package-lock` and do a clean install to let `npm` resolve dependencies properly.
 
> Btw, could you install latest version of `@smg-automotive/eslint-config`, I have just added the [`eslint-import-resolver-typescript`](https://www.npmjs.com/package/eslint-import-resolver-typescript) to resolve all custom path defined in `tsconfig.json`

Done, thanks 👍🏻

**Comment by Averethel** [comment]
> > @quangpdt can you take a look at the configuration of import/no-internal-modules? It currently prevents me from doing relative imports
> 
> @Averethel. I understand the problem. From perspective of `import/no-internal-modules`, each child directory of `src` should serve for a single purpose and all public modules should be exported as a single point, `index.ts`. Personally, I found it's quite too strict to use for development and only should be applied when importing from Dependencies, there's still [open discussion](https://github.com/import-js/eslint-plugin-import/issues/1430) regarding this topic.
> 
> To get over this issue, I would recommend to update `tsconfig.json` and your `.eslintrc.js` for minor customization:
> 
> ```json
> {
>   "compilerOptions": {
>     "baseUrl": "./src",
>     "paths": {
>       "~/*": ["./*"]
>     }
>   },
> }
> ```
> 
> Then, allow internal modules could be imported within our app scope in `.eslintrc.js` with custom rule:
> 
> ```js
> 'import/no-internal-modules': [
>   'error',
>   {
>     allow: ['~/**/*'],
>   },
> ],
> ```
> 
> For Next.js application, we'll need to do some more tweaks, but if we all agree with this, I can re-adapt to the ESLint config so you don't need to customize the `.eslintrc.js` anymore. My idea is updating rule `import/no-internal-modules` with some exceptions that can cover all kind of projects (TypeScript, React, Next.js)
> 
> ```js
> 'import/no-internal-modules': [
>   'error',
>   {
>     allow: ['~/**/*', '@pages/**/*', , '@components/**/*'],
>   },
> ],
> ```

I made it work with `index` files in every directory but it is a bit of overhead. While I like tilde imports one thing that makes them a bit annoying is that the whole configuration needs to be repeated in jest config. And it still doesn't fully solve the problem because it would favor:
```
import x from '~/something/very/deeply/nested/oh/my/god'
```
over:
```
import x from '../god'
```
could we allow nested modules for relative import? 
And let's see if we still need `~` imports in `next` projects afterwards?

**Comment by quangpdt** [comment]
Unfortunately, I cannot find any way that we can allow both nested relative import path and `import/no-internal-modules` in parallel. If you found something, please share your thought.

At the moment, I only can give some proposals:
- If you found the absolute import path is too long, you can just create a new alias path in `tsconfig.json`, and add it as a new exception in `eslintrc.js`, like I described above. In TS world, the alias path is very widely used and also can minimize the import path length, help us easier to read the purpose of that module.
- Turn off this rules temporarily until we find better way to resolve the problem.

**Comment by Averethel** [comment]
All right, thanks for checking. Let's roll with it as is. This is by no means a blocker and we can tweak and improve it later
